### ISSUE TITLE
Increase required C++ version to C++17

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#3918]
**New features**

* constexpr if, constexpr lambda
* std::optional, std::filesystem
* template argument deduction
  ```c++
  #include <array>
  std::array a = {1,2,3};
  ```
* structured bindings and tuple unpacking
  ```c++
  #include <tuple>
  inline auto calc_values(int flag) {
    std::array f1{1, 2, 3}, f2{4, 5, 6};
    if (flag)
      return std::tuple{f1, f2};
    else
      return std::tuple{f2, f1};
  }
  int get_values() {
    auto [f1, f2] = calc_values(1);
    return f1[0];
  }
  ```
* fold-expressions
  ```c++
  template<typename ...Args>
  int sum_plus_three(Args&&... args) {
    return (args + ... + 3);
  }
  int foo() {
    return sum_plus_three(0, 1, 2); // returns 6
  }
  ```

**Prior work**

ESPResSo is C++17-ready since #3824, including in the walberla branch.

**Course of action**

* Update `CMAKE_CXX_STANDARD` to 17
* Sort out the failing exit function check in the CI build script ([build_cmake.sh#L215-L224](https://github.com/espressomd/espresso/blob/f760050199b94226b9127b309bfa7db569507605/maintainer/CI/build_cmake.sh#L215-L224))
* Keep CUDA 11. Drop support for CUDA 8, 9, 10: the nvcc compiler doesn't support C++17 in these versions.

Clang can be used as a C++17 CUDA compiler for older CUDA versions, e.g. Clang 6 can compile ESPResSo with CUDA 8 and C++17-specific code in .cu files. However Clang and CUDA have a complicated inter-dependency graph, e.g. CUDA 10 requires Clang 8 or lower, but Clang 8 doesn't support the CUDA 10 library. I don't think sorting out compatible Clang versions is worth it.

**Comment by mkuron** [comment]
According to https://en.cppreference.com/w/cpp/compiler_support, CUDA 11's C++17 support is quite incomplete, so it's probably a bit early to make the switch. Cray does not support C++17 yet, but since Hazel Hen has been replaced with Hawk, nobody should care. Intel should have pretty good C++17 support by now, but we'll need to keep an eye on it since it's not in CI anymore.

> Drop support for CUDA 8, 9, 10

That's going to be a problem for our cluster and some of our workstation. Support for Kepler chips has been dropped in CUDA 11.

> However Clang and CUDA have a complicated inter-dependency graph, e.g. CUDA 10 requires Clang 8 or lower, but Clang 8 doesn't support the CUDA 10 library. I don't think sorting out compatible Clang versions is worth it.

It sounds like you are mixing two dependency graphs here:
1. One is for using nvcc with clang as the host compiler, and that's documented here: https://gist.github.com/ax3l/9489132#nvcc. To get C++17 support with this toolchain, you need CUDA 11 and Clang 6-9.
2. One is for using clang as a single-pass compiler for both device and host, which only uses ptxas from the CUDA toolkit. To get C++17 support with this toolchain, you need a Clang version that supports C++17 (which according to https://en.cppreference.com/w/cpp/compiler_support would be Clang 5-11) and the installed CUDA toolkit (which according to https://gist.github.com/ax3l/9489132#clang--x-cuda can be as low as 7.0).

Since compiler error messages about incompatible versions are pretty good, we don't need any CMake-level version checks (and we don't currently have them anyway). For the clang CI job, we could either upgrade to Clang 10 with CUDA 10.1 (the latest supported combination), stick to our current combination of Clang 9 with CUDA 10.1, or move to Clang 11 with CUDA 11 once LLVM 11 is released in a few weeks.

**Comment by jngrad** [comment]
> CUDA 11's C++17 support is quite incomplete

Thanks for posting this chart, I didn't realize how few of the new features were supported by nvcc. `std::optional` isn't supported, which would have been nice for the LB GPU thermostat. This was the only change I had in mind for the CUDA code though, so it's not a deal breaker. My main motivation here is to get C++17 in the C++ core. We can probably live with a C++14 restriction on the subset of header files that are visible to the LB GPU code (the hassle wouldn't be worse than the current MPI include restriction), until we remove LB GPU in release 4.3.

> It sounds like you are mixing two dependency graphs here

I didn't put too much thought into sorting out Clang-CUDA compatibility. Some experimenting with Clang 6 and 9 has shown me that it was a lot of effort to figure out which combinations worked and which didn't. I would be in favor of supporting a Clang-CUDA11 combination that works out of the box, so that we can put a one-liner Clang version check in the CMake logic in GPU builds to avoid spending time looking at compatibility charts.

> That's going to be a problem for our cluster and some of our workstation. Support for Kepler chips has been dropped in CUDA 11.

The future of CUDA in the 4.3 release is uncertain, in any case. People who rely on GPU acceleration can still use the 4.2 release, which supports CUDA 8-11 and will get bugfixes until October 2021.

**Comment by RudolfWeeber** [comment]
Postponed 4.3 due to Cuda 10 shortcomings

**Comment by jngrad** [comment]
Support for CUDA 8 and 9 has been dropped in 4.2.

Regarding C++17 and our build system:
- function `Utils::apply()` needs to be removed (superseded by `std::apply()` introduced in C++17, causes a compiler warning due to the ambiguous call)
- GPU support leaves us with two options to compile in C++17 mode:
   - compile with CUDA 10 and Clang 10, by adding `-fsized-deallocation` and `-UNDEBUG`, see details below
   - compile with CUDA 11 and GCC 9 + NVCC 11
   - other compiler/cuda combinations are problematic (#4081)

Clang 10 workarounds:
- `-fsized-deallocation` fixes the following `thrust` error
   ```
   /usr/lib/llvm-10/lib/clang/10.0.0/include/cuda_wrappers/thrust/mr/new.h:67:9: error: no matching function for call to 'operator delete' [clang-diagnostic-error]
           ::operator delete(p, bytes, std::align_val_t(alignment));
           ^
   ```
- `-UNDEBUG` re-enables assertions to bypass a Clang-Tidy alert that cannot be silenced by `// NOLINT(clang-analyzer-core.NullDereference)`; try the following MWE with `clang-tidy-10 mwe.cpp -- -isystem /usr/lib/x86_64-linux-gnu/openmpi/include -DNDEBUG -pthread -std=c++17` (either use c++14 or add `-UNDEBUG` to compile successfully)
   ```c++
   #include <boost/mpi.hpp>
   #include <boost/mpi/communicator.hpp>
   #include <boost/mpi/nonblocking.hpp>
   #include <vector>

   int main(int argc, char **argv) {
     boost::mpi::environment mpi_env(argc, argv);
     try {
       std::vector<boost::mpi::request> req = {};
       boost::mpi::wait_all(req.begin(), req.end());
     } catch (...) { }
   }
   ```

**Comment by RudolfWeeber** [comment]
We should probably direct users to compile with gcc + nvcc. 

Having one clang-base version would be good for devs, due to better error diagnostics and due to static analysis and sanitizers.

I think we should look into clang-11, directly, considering that 4.3 will be released late this year or after.

**Comment by mkuron** [comment]
> GPU support leaves us with two options to compile in C++17 mode

In Walberla, we were recently [faced with the same problem](https://i10git.cs.fau.de/walberla/walberla/-/merge_requests/373). As long as you keep C++17 language features out of the header files included by .cu files, it's not difficult though. You can even use C++17 library features like `std::optional`, `std::any` and `std::filesystem` because stdlibc++ provides C++14-compatible versions in `std::experimental` -- just be sure to detect them on the CMake side so you don't end up mixing `std` and `std::experimental` across different compilation units.

**Comment by jngrad** [comment]
nvcc 11.0 now supports most of the C++17 core language features ([cppreference chart](https://en.cppreference.com/w/cpp/compiler_support#C.2B.2B17_features)). We need nvcc 10.1 support for the ICP workstations and cluster, so we have to keep supporting C++14 in 4.2. Since waLBerla now requires C++17, 4.2 needs C++17 support too because it is developed in parallel to 4.3. This is trivial, we only need to change 4 lines of code in a unit test to fix the `apply()` error mentioned earlier and 4.2 will support both C++ standards. C++17 support can be tested on `maxset` which doesn't need a GPU.
